lib/semantics/expression.cc

```diff
@@ -915,6 +918,16 @@ static std::optional<Component> CreateComponent(
   return std::nullopt;
 }
 
+static const semantics::DerivedTypeSpec *GetDerivedTypeSpec(
```

Comment:
I think this function should be in lib/evaluate/type.{h,cc}, and some calls to `GetDerivedTypeSpec()` in the codebase that are protected by similar predicates could be replaced with calls to this utility.  Also, I think you should squash the predicates together, they don't have to be distinct `if` statements, but that's just my preference.

Reply:
Good idea, this simplifies the code in a bunch of places.